components/places/src/match_impl.rs

```diff
@@ -0,0 +1,483 @@
+
+/// Returns true if `source` starts with `token` ignoring case.
+///
+/// Loose port of stringMatch from places, that does more correct case folding.
+/// (If this turns out to be a perf issue we can always address it then)
+#[inline]
+fn string_match(token: &str, source: &str) -> bool {
+    if source.len() < token.len() {
+        return false;
+    }
+    let mut ti = token.chars().default_case_fold();
+    let mut si = source.chars().default_case_fold();
+    let mut i = 0;
+    loop {
+        match (ti.next(), si.next()) {
+            (None, _) => return true,
+            (Some(_), None) => return false,
+            (Some(x), Some(y)) => if x != y { return false },
+        }
+        i += 1;
+        if i > token.len() + source.len() {
+            panic!("Bug: {:?} {:?}", token, source);
+        }
+    }
+}
+
+/// This performs single-codepoint case folding. It will do the wrong thing
+/// for characters which .
+#[inline]
+fn char_to_lower_single(c: char) -> char {
+    c.to_lowercase().next().unwrap()
+}
+
+#[inline]
+fn next_codepoint_lower(s: &str) -> (usize, char) {
+    // This is super convoluted, and I wish a more direct way to do it was exposed
+    let mut indices = s.char_indices();
+    let (_, next_char) = indices.next().unwrap();
+    let next_index = if let Some((index, _)) = indices.next() { index } else { s.len() };
+    (next_index, char_to_lower_single(next_char))
+}
+
+// Port of places `findInString`.
+pub fn find_in_string(token: &str, src: &str, only_boundary: bool) -> bool {
+    // Place's verison has this restriction too
+    assert!(!token.is_empty(), "Don't search for an empty string");
+    if src.len() < token.len() {
+        return false;
+    }
+
+    let token_first_char = next_codepoint_lower(token).1;
+    // The C++ code is a big ol pointer party, and even indexes with negative numbers
+    // in some places. We aren't quite this depraved, so we just use indices into slices.
+    //
+    // There's probably a higher cost to this than usual, and if we had more robust testing
+    // (fuzzing, even) it might be worth measuring a version of this that avoids more of the
+    // bounds checks.
+    let mut cur_offset = 0;
+    // Scan forward to the next viable candidate (if any).
+    while let Some(src_idx) = next_search_candidate(&src[cur_offset..], token_first_char) {
+        if src_idx >= src.len() {
+            break;
+        }
+        cur_offset += src_idx;
+        let src_cur = &src[cur_offset..];
+
+        // Check whether the first character in the token matches the character
+        // at sourceCur. At the same time, get the index of the next character
+        // in the source.
+        let (next_offset_in_cur, src_next_char) = next_codepoint_lower(src_cur);
+
+        // If it is the first character, and we either don't care about boundaries or
+        // we're on one, do the more expensive string matching and return true if it hits.
+        if src_next_char == token_first_char
```

Comment:
So even if we have a string like `aaa` and we're searching for `a`, we'll still bail thanks to `cur_offset == 0`, right?

Reply:
Basically, I've tried to make it a bit more clear. the `cur_offset == 0` check is just because it was illegal to call is_on_boundary with cur_offset as 0